### ISSUE TITLE
Custom loss function explodes and models won't learn

### ISSUE THREAD

**Issue by Meight** [qubvel/segmentation_models#31]
Thanks for the great repository and work, this is so helpful!

Recently I've trained a TensorFlow-based ResNet101 on Pascal VOC 2012 dataset. Its particularity is that there's 21 plus one class used to denote "ambiguous" regions, meant to be ignored when computing the loss and metrics. Pixels marked as ambiguous have value 255.

For this purpose, I have defined a custom loss function that I'm successfully using in the TF-based version, as follow:

```python
def restricted_categorical_crossentropy(y_true, y_pred, class_labels):
    def get_valid_probabilities_and_labels(annotation_batch_tensor,
                                    logits_batch_tensor,
                                    class_labels):
        valid_labels_mask = tf.not_equal(annotation_batch_tensor,
                                         mask_out_class_label) # in this case mask_out_class_label is equal to 255.
        valid_labels_indices = tf.where(valid_labels_mask)
        valid_batch_indices = tf.to_int32(valid_labels_indices)

        valid_labels_batch_tensor = tf.gather_nd(params=annotation_batch_tensor, indices=valid_batch_indices)
        valid_probabilities_batch_tensor = tf.gather_nd(params=logits_batch_tensor, indices=valid_batch_indices)

        return valid_labels_batch_tensor, valid_probabilities_batch_tensor

    valid_labels, valid_predictions = get_valid_probabilities_and_labels(
        annotation_batch_tensor=y_true,
        logits_batch_tensor=y_pred,
        class_labels=class_labels)

    cross_entropy = K.categorical_crossentropy(valid_labels, valid_predictions)

    # The number of elements is different during each step due to mask out regions: normalize entropy.
    return K.mean(cross_entropy)
```

When I compile any of this repo's Keras models using Keras' native `K.categorical_crossentropy`, I get fairly decent values starting at around 2.5.

However, when I compile it using my custom loss function, the values explode to things like 11563889.6988.

How does this make any sense granted

1. my custom function uses `K.categorical_crossentropy` on a subset of pixels of the same batches;
2. that exact same function works well when I'm not using Keras (I get values close to 2.5 and lower)?

Of course I'm using the exact same preprocessing functions and the batching process is literally copy/pasted so I doubt it's coming from the data.

For both loss functions (`K.categorical_crossentropy` or `restricted_categorical_crossentropy`) none of the models I've tried compiling from this repo have been able to learn anything worthwhile when my TF-based implementation could. The accuracy increases to about 84% but the mean IoU is ridiculously low (around 4% which for 21 classes is nothing more than random assignment.) In comparison, my TF-based model could reach around 55% mIoU, when also pretrained on ImageNet.

Do you have any thought on where my mistake(s) could be by any chance @qubvel ?

Thanks in advance for your kind help!
